File: app/src/services/assignmentService.js
import apiClient from './apiClient';
import { useAuthStore } from '../store/authStore';

/**
 * Assignment Service - Build assignment management
 * Handles two-layer access control: RBAC + explicit build assignments
 */
class AssignmentService {
  /**
   * Create a build assignment
   * @param {string} buildId - Build ID
   * @param {string} userId - User ID to assign
   * @param {string} personaRole - Persona role (workload_owner, data_owner, auditor)
   * @returns {Promise<Object>}
   */
  async createAssignment(buildId, userId, personaRole) {
    const response = await apiClient.post(`/builds/${buildId}/assignments`, {
      user_id: userId,
      persona_role: personaRole
    });

    return response.data;
  }

  /**
   * Get all assignments for a build
   * @param {string} buildId - Build ID
   * @returns {Promise<Array>}
   */
  async getBuildAssignments(buildId) {
    const response = await apiClient.get(`/builds/${buildId}/assignments`);
    return response.data.assignments || [];
  }

  /**
   * Get all assignments for a user
   * @param {string} userId - User ID
   * @returns {Promise<Array>}
   */
  async getUserAssignments(userId) {
    const response = await apiClient.get(`/users/${userId}/assignments`);
    return response.data.assignments || [];
  }

  /**
   * Get current user's assignments
   * @returns {Promise<Array>}
   */
  async getMyAssignments() {
    const user = useAuthStore.getState().user;
    if (!user) throw new Error('User not authenticated');

    return this.getUserAssignments(user.id);
  }

  /**
   * Delete a specific assignment
   * @param {string} buildId - Build ID
   * @param {string} userId - User ID
   * @param {string} personaRole - Persona role
   * @returns {Promise<void>}
   */
  async deleteAssignment(buildId, userId, personaRole) {
    await apiClient.delete(`/builds/${buildId}/assignments`, {
      data: {
        user_id: userId,
        persona_role: personaRole
      }
    });
  }

  /**
   * Delete all assignments for a persona role in a build
   * @param {string} buildId - Build ID
   * @param {string} personaRole - Persona role
   * @returns {Promise<void>}
   */
  async deleteBuildAssignments(buildId, personaRole) {
    await apiClient.delete(`/builds/${buildId}/assignments`, {
      data: { persona_role: personaRole }
    });
  }

  /**
   * Validate if user is assigned to a build for a specific role
   * @param {string} buildId - Build ID
   * @param {string} userId - User ID
   * @param {string} personaRole - Persona role
   * @returns {Promise<boolean>}
   */
  async validateUserAssignment(buildId, userId, personaRole) {
    try {
      const assignments = await this.getBuildAssignments(buildId);
      return assignments.some(a =>
        a.user_id === userId && a.persona_role === personaRole
      );
    } catch (error) {
      console.error('Failed to validate assignment:', error);
      return false;
    }
  }

  /**
   * Check if current user is assigned to a build for a specific role
   * @param {string} buildId - Build ID
   * @param {string} personaRole - Persona role
   * @returns {Promise<boolean>}
   */
  async isAssignedToRole(buildId, personaRole) {
    const user = useAuthStore.getState().user;
    if (!user) return false;

    return this.validateUserAssignment(buildId, user.id, personaRole);
  }

  /**
   * Check if current user can submit a section
   * Validates assignment and checks if section already submitted
   * @param {string} buildId - Build ID
   * @param {string} personaRole - Persona role
   * @param {Array} existingSections - Existing sections for the build
   * @returns {Promise<{canSubmit: boolean, reason: string}>}
   */
  async canSubmitSection(buildId, personaRole, existingSections = []) {
    const user = useAuthStore.getState().user;
    if (!user) {
      return { canSubmit: false, reason: 'User not authenticated' };
    }

    // Check assignment
    const isAssigned = await this.validateUserAssignment(buildId, user.id, personaRole);
    if (!isAssigned) {
      return {
        canSubmit: false,
        reason: 'You are not assigned to this build for this role'
      };
    }

    // Check if section already submitted
    const alreadySubmitted = existingSections.some(s => s.persona_role === personaRole);
    if (alreadySubmitted) {
      return {
        canSubmit: false,
        reason: 'Section already submitted for this role'
      };
    }

    return { canSubmit: true, reason: '' };
  }

  /**
   * Get assignment summary for a build
   * @param {string} buildId - Build ID
   * @returns {Promise<Object>} - {total, byRole: {workload_owner: [], data_owner: [], auditor: []}}
   */
  async getAssignmentSummary(buildId) {
    const assignments = await this.getBuildAssignments(buildId);

    const byRole = {
      workload_owner: [],
      data_owner: [],
      auditor: []
    };

    assignments.forEach(assignment => {
      if (byRole[assignment.persona_role]) {
        byRole[assignment.persona_role].push(assignment);
      }
    });

    return {
      total: assignments.length,
      byRole,
      workloadOwnerCount: byRole.workload_owner.length,
      dataOwnerCount: byRole.data_owner.length,
      auditorCount: byRole.auditor.length
    };
  }

  /**
   * Check if all required roles are assigned
   * @param {string} buildId - Build ID
   * @returns {Promise<{complete: boolean, missing: Array<string>}>}
   */
  async checkAssignmentCompleteness(buildId) {
    const summary = await this.getAssignmentSummary(buildId);
    const missing = [];

    if (summary.workloadOwnerCount === 0) missing.push('workload_owner');
    if (summary.dataOwnerCount === 0) missing.push('data_owner');
    if (summary.auditorCount === 0) missing.push('auditor');

    return {
      complete: missing.length === 0,
      missing
    };
  }
}

export default new AssignmentService();

File: backend/internal/service/assignment_service.go
package service

import (
	"context"
	"fmt"
	"log/slog"
	"strings"

	"github.com/google/uuid"

	"github.com/Sashwat-K/persona-based-contract-generator/backend/internal/model"
	"github.com/Sashwat-K/persona-based-contract-generator/backend/internal/repository"
)

// AssignmentService handles build-to-user-to-role assignments.
type AssignmentService struct {
	queries      repository.Querier
	auditService *AuditService
}

// NewAssignmentService creates a new AssignmentService.
func NewAssignmentService(queries repository.Querier, auditService *AuditService) *AssignmentService {
	return &AssignmentService{
		queries:      queries,
		auditService: auditService,
	}
}

// CreateAssignmentInput contains data for creating a build assignment.
type CreateAssignmentInput struct {
	BuildID    uuid.UUID
	RoleID     *uuid.UUID
	RoleName   string // Backward-compatible
	UserID     uuid.UUID
	AssignedBy uuid.UUID
	IPAddress  string
	Signature  *string
	SigHash    *string
}

// CreateAssignment assigns a user to a specific role for a build.
// This implements the two-layer access control: role + explicit assignment.
func (s *AssignmentService) CreateAssignment(ctx context.Context, input CreateAssignmentInput) (*repository.BuildAssignment, error) {
	// 1. Validate build exists and is not finalized
	build, err := s.queries.GetBuildByID(ctx, input.BuildID)
	if err != nil {
		if strings.Contains(err.Error(), "no rows") {
			return nil, model.ErrNotFound("build not found")
		}
		return nil, fmt.Errorf("failed to get build: %w", err)
	}

	buildStatus := model.BuildStatus(build.Status)
	if buildStatus.IsTerminal() {
		return nil, model.ErrInvalidRequest(fmt.Sprintf("cannot assign users to %s build", buildStatus))
	}

	// 2. Validate role exists
	var role repository.Role
	if input.RoleID != nil && *input.RoleID != uuid.Nil {
		role, err = s.queries.GetRoleByID(ctx, *input.RoleID)
		if err != nil {
			if strings.Contains(err.Error(), "no rows") {
				return nil, model.ErrNotFound("role not found")
			}
			return nil, fmt.Errorf("failed to get role: %w", err)
		}
	} else {
		role, err = s.queries.GetRoleByName(ctx, input.RoleName)
		if err != nil {
			if strings.Contains(err.Error(), "no rows") {
				return nil, model.ErrNotFound(fmt.Sprintf("role '%s' not found", input.RoleName))
			}
			return nil, fmt.Errorf("failed to get role: %w", err)
		}
	}

	// 3. Validate user exists
	_, err = s.queries.GetUserByID(ctx, input.UserID)
	if err != nil {
		if strings.Contains(err.Error(), "no rows") {
			return nil, model.ErrNotFound("user not found")
		}
		return nil, fmt.Errorf("failed to get user: %w", err)
	}

	// 4. Check if this specific user is already assigned to this role for this build
	alreadyAssigned, err := s.queries.CheckUserAssignedToBuild(ctx, repository.CheckUserAssignedToBuildParams{
		BuildID: input.BuildID,
		UserID:  input.UserID,
		RoleID:  role.ID,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to check existing assignment: %w", err)
	}
	if alreadyAssigned {
		return nil, model.ErrInvalidRequest(fmt.Sprintf("user is already assigned to role '%s' for this build", input.RoleName))
	}

	// 5. Create the assignment
	assignment, err := s.queries.CreateBuildAssignment(ctx, repository.CreateBuildAssignmentParams{
		BuildID:    input.BuildID,
		RoleID:     role.ID,
		UserID:     input.UserID,
		AssignedBy: input.AssignedBy,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create assignment: %w", err)
	}

	// 6. Audit log the assignment
	eventData := map[string]string{
		"assignment_id": assignment.ID.String(),
		"role_name":     role.Name,
		"role_id":       role.ID.String(),
		"user_id":       input.UserID.String(),
	}
	if input.SigHash != nil && *input.SigHash != "" {
		eventData["request_signature_hash"] = *input.SigHash
	}

	_, err = s.auditService.LogEvent(ctx, LogEventInput{
		BuildID:     input.BuildID,
		EventType:   model.EventRoleAssigned,
		ActorUserID: input.AssignedBy,
		IpAddress:   input.IPAddress,
		EventData:   eventData,
		Signature:   input.Signature,
	})
	if err != nil {
		// Best effort audit logging: assignment should still succeed.
		slog.Error("failed to log assignment audit event",
			"error", err,
			"assignment_id", assignment.ID,
			"build_id", input.BuildID,
			"role_id", role.ID,
			"user_id", input.UserID,
		)
	}

	return &assignment, nil
}

// GetBuildAssignments returns all assignments for a build with enriched data.
func (s *AssignmentService) GetBuildAssignments(ctx context.Context, buildID uuid.UUID) ([]repository.GetBuildAssignmentsByBuildIDRow, error) {
	assignments, err := s.queries.GetBuildAssignmentsByBuildID(ctx, buildID)
	if err != nil {
		return nil, fmt.Errorf("failed to get build assignments: %w", err)
	}
	return assignments, nil
}

// GetUserAssignments returns all builds a user is assigned to.
func (s *AssignmentService) GetUserAssignments(ctx context.Context, userID uuid.UUID) ([]repository.GetBuildAssignmentsByUserIDRow, error) {
	assignments, err := s.queries.GetBuildAssignmentsByUserID(ctx, userID)
	if err != nil {
		return nil, fmt.Errorf("failed to get user assignments: %w", err)
	}
	return assignments, nil
}

// CheckUserAssignment verifies if a user is assigned to a specific role for a build.
func (s *AssignmentService) CheckUserAssignment(ctx context.Context, buildID, userID uuid.UUID, roleName string) (bool, error) {
	// Get role ID from name
	role, err := s.queries.GetRoleByName(ctx, roleName)
	if err != nil {
		if strings.Contains(err.Error(), "no rows") {
			return false, nil // Role doesn't exist, so user can't be assigned
		}
		return false, fmt.Errorf("failed to get role: %w", err)
	}

	isAssigned, err := s.queries.CheckUserAssignedToBuild(ctx, repository.CheckUserAssignedToBuildParams{
		BuildID: buildID,
		UserID:  userID,
		RoleID:  role.ID,
	})
	if err != nil {
		return false, fmt.Errorf("failed to check assignment: %w", err)
	}

	return isAssigned, nil
}

// DeleteBuildAssignments removes all assignments for a build (used when cancelling).
func (s *AssignmentService) DeleteBuildAssignments(ctx context.Context, buildID uuid.UUID, actorID uuid.UUID, ip string) error {
	// Verify build exists
	build, err := s.queries.GetBuildByID(ctx, buildID)
	if err != nil {
		if strings.Contains(err.Error(), "no rows") {
			return model.ErrNotFound("build not found")
		}
		return fmt.Errorf("failed to get build: %w", err)
	}

	buildStatus := model.BuildStatus(build.Status)
	if buildStatus.IsTerminal() {
		return model.ErrInvalidRequest(fmt.Sprintf("cannot delete assignments from %s build", buildStatus))
	}

	// Delete assignments
	err = s.queries.DeleteBuildAssignmentsByBuildID(ctx, buildID)
	if err != nil {
		return fmt.Errorf("failed to delete assignments: %w", err)
	}

	// Audit log
	_, err = s.auditService.LogEvent(ctx, LogEventInput{
		BuildID:     buildID,
		EventType:   model.EventRoleAssigned, // Reuse event type with negative context
		ActorUserID: actorID,
		IpAddress:   ip,
		EventData: map[string]string{
			"action": "delete_all_assignments",
		},
	})
	if err != nil {
		return fmt.Errorf("assignments deleted but failed to log audit event: %w", err)
	}

	return nil
}

// ValidateAssignmentForSubmission checks if a user can submit a section for a role.
// This enforces the two-layer access control.
func (s *AssignmentService) ValidateAssignmentForSubmission(ctx context.Context, buildID, userID uuid.UUID, roleName string) error {
	// Layer 1: RBAC (user must have the role globally).
	userRoles, err := s.queries.GetRolesByUserID(ctx, userID)
	if err != nil {
		return fmt.Errorf("failed to load user roles: %w", err)
	}
	hasRole := false
	for _, r := range userRoles {
		if r.Role == roleName {
			hasRole = true
			break
		}
	}
	if !hasRole {
		return model.ErrForbidden(fmt.Sprintf("user does not have role '%s'", roleName))
	}

	// Layer 2: Explicit build assignment.
	isAssigned, err := s.CheckUserAssignment(ctx, buildID, userID, roleName)
	if err != nil {
		return fmt.Errorf("failed to check assignment: %w", err)
	}

	if !isAssigned {
		return model.ErrForbidden(fmt.Sprintf("user is not assigned to role '%s' for this build", roleName))
	}

	return nil
}
